### ISSUE TITLE
ASP.NET Core project does not start when loaded into a separate AssemblyLoadContext

### ISSUE THREAD

**Issue by bryanenroute** [aspnet/AspNetCore#17098]
Here's a link to the code to reproduce the issue:  https://github.com/bryanenroute/aspnetcore-alc-issue

I am loading an ASP.NET Core project into a separate custom AssemblyLoadContext using a plugin style implementation where the main application (NetCoreConsoleApp) and the ASP.NET Core project (AspNetCoreApp) reference a common interface (IModule, NetStandardCommon .NET Standard project).

Although the plugin/module loads, the console displays as follows:

`info: Microsoft.Hosting.Lifetime[0]
      Now listening on: http://localhost:5000
info: Microsoft.Hosting.Lifetime[0]
      Now listening on: https://localhost:5001
crit: Microsoft.AspNetCore.Hosting.Diagnostics[11]
      Hosting startup assembly exception
System.InvalidOperationException: Startup assembly AspNetCoreApp failed to execute. See the inner exception for more details.
 ---> System.IO.FileNotFoundException: Could not load file or assembly 'AspNetCoreApp, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.
File name: 'AspNetCoreApp, Culture=neutral, PublicKeyToken=null'
   at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, RuntimeAssembly assemblyContext, StackCrawlMark& stackMark, Boolean throwOnFileNotFound, AssemblyLoadContext assemblyLoadContext)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, StackCrawlMark& stackMark, AssemblyLoadContext assemblyLoadContext)
   at System.Reflection.Assembly.Load(AssemblyName assemblyRef, StackCrawlMark& stackMark, AssemblyLoadContext assemblyLoadContext)
   at System.Reflection.Assembly.Load(AssemblyName assemblyRef)
   at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.ExecuteHostingStartups()

   --- End of inner exception stack trace ---
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
info: Microsoft.Hosting.Lifetime[0]
      Content root path: X:\Github\aspnetcore-alc-issue\NetCoreConsoleApp\bin\Debug\netcoreapp3.0`

While this seems to work for 'Hello World', if you add anything meaningful to Startup.ConfigureServices (like services.AddRazorPages()), an exception is thrown:

`{"Could not load file or assembly 'AspNetCoreApp, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.":"AspNetCoreApp, Culture=neutral, PublicKeyToken=null"}`

This is related to another issue I submitted (https://github.com/dotnet/core-setup/issues/8682).

**Update:  Changed AspNetCoreApp code to follow SignalR tutorial:  https://docs.microsoft.com/en-us/aspnet/core/tutorials/signalr?view=aspnetcore-3.0&tabs=visual-studio

If you you run AspNetCoreApp directly, the chat example works.  If you run NetCoreConsoleApp, it fails to load AspNetCoreApp.**

@vitek-karas

**Comment by davidfowl** [comment]
TL;DR Add this line of code:

```C#
using var _ = context.EnterContextualReflection();
```

Full code sample:

```C#
using NetStandardCommon;
using System;
using System.IO;

namespace NetCoreConsoleApp
{
    class Program
    {
        static void Main(string[] args)
        {
            //Fails!
            FileInfo asm = new FileInfo(@"..\..\..\..\AspNetCoreApp\bin\debug\netcoreapp3.0\AspNetCoreApp.dll");
            var moduleDirectory = asm.DirectoryName;

            ModuleAssemblyLoadContext context = new ModuleAssemblyLoadContext(asm.Name, moduleDirectory, typeof(IModule));
            context.Scan();

            using var _ = context.EnterContextualReflection();

            foreach (var module in context.GetImplementations<IModule>())
            {
                module.Start();
            }
        }
    }
}
```

Reasoning:

First, a couple of things:
- Everything works fine, but a failure is logged because `Assembly.Load(Application)` doesn't work
- This isn't specific to ASP.NET Core, the issue occurs when you do ModuleHost -> Module -> Library -> Assembly.Load(Module). IIRC Assembly.Load uses some magic to figure out which `AssemblyLoadContext` should be used and it can't figure it out in this case. Calling `EnterContextualReflection` makes the `ModuleAssemblyLoadContext` the "right one".

**Comment by bryanenroute** [comment]
@davidfowl Excellent!  With that line, it works and serves the page.  I updated the sample code to use the SignalR tutorial... the page loads, but the 'chat' function doesn't seem to work.  Just starting to investigate, but if there's something easy I'm missing let me know.

**Comment by bryanenroute** [comment]
@davidfowl Thank you, invoking EnterContextualReflection did the trick!  The only other tweak I made was changing the web root:

`    public class Program
    {
        public static void Main(string[] args)
        {
            CreateHostBuilder(args).Build().Run();
        }

        public static IHostBuilder CreateHostBuilder(string[] args) =>
            Host.CreateDefaultBuilder(args)
                .ConfigureWebHostDefaults(webBuilder =>
                {
                    webBuilder.UseWebRoot(@"..\..\..\..\AspNetCoreApp\wwwroot")
                    .UseStartup<Startup>();
                });
    }`
